Replace Honcho with Supervisor

seanh opened the pull request:
Same as we've done for lms and h:

https://github.com/hypothesis/lms/pull/1897
https://github.com/hypothesis/h/pull/6103

This seems to have been working well for h and lms: `make dev` is stopping quickly and reliably when Ctrl-C'd.

Via 3 has still been having the old problems and is now inconsistent with h and lms so let's bring it into line.

This is the last app to remove Honcho from: bouncer and legacy Via don't use Honcho (`make dev` only runs one process)

seanh committed:
Replace Honcho with Supervisor

Same as we've done for lms and h:

https://github.com/hypothesis/lms/pull/1897
https://github.com/hypothesis/h/pull/6103

This seems to have been working well for h and lms: `make dev` is
stopping quickly and reliably when Ctrl-C'd.

Via 3 has still been having the old problems and is now inconsistent
with h and lms so let's bring it into line.

seanh commented on bin/logger:
```diff
@@ -64,24 +64,69 @@ we redirect the aggregated output back to STDOUT (FD 1). You can also log the
 aggregated output to a single file.
 """
 
+import itertools
```
This is just a copy-paste of the entire new version of the `bin/logger` file from h and lms

seanh commented on conf/supervisord-dev.conf:
```diff
@@ -0,0 +1,35 @@
+stopasgroup = true
+
+[program:nginx]
+command=docker-compose up --no-color --exit-code-from nginx-proxy nginx-proxy
```
This is slightly more complicated than just `docker-compose up`:

1. We only want to run the NGINX container (named `nginx-proxy`), and not any other service containers that might be defined in `docker-compose.yml` in future. Hence `docker-compose up ... nginx-proxy' (in fact `nginx-proxy` is the only container in `docker-compose.yml at the moment). To start just one container and also print its output it seems that you need to do `docker-compose up -

2. Added `--no-color` to let Supervisor (specifically: our `bin/logger` script) colorize the output not docker-compose

3 . Added `--exit-code-from` to tell docker-compose to proxy nginx's exit code

There are a couple of problems with this that're caused by using docker-compose to run nginx instead of running docker directly. These are pre-existing problems, for another PR:

1. Nginx is still left running after you <kbd><kbd>Ctrl</kbd> + <kbd>c</kbd></kbd> the `make dev` process. This isn't the right behaviour for `make dev` which should kill all the processes when stopped (and does so for every other process across our apps, except this one)

2. When you run `make dev` it starts by printing out a potentially long number of log lines from nginx from previous runs of `nginx`

Separate issue: https://github.com/hypothesis/via3/issues/221

jon-betts replied:
You probably want to take the same approach as `superdev` which has solved the closing on exit issue: 

* https://github.com/hypothesis/superdev/blob/master/conf/via3/services.conf

I would strongly suggest checking `superdev` as a first resort when attempting to work out how to run these.

robertknight replied:
There is a `docker-compose run` command which can run a docker-compose service in a new container. With the following change I got `make dev` to start an nginx container which was then killed when `make dev` was interrupted. Making this second part work required changing the signal.

```diff
diff --git a/conf/supervisord-dev.conf b/conf/supervisord-dev.conf
index 12622bb..bb5cd8e 100644
--- a/conf/supervisord-dev.conf
+++ b/conf/supervisord-dev.conf
@@ -10,10 +10,10 @@ stopsignal = KILL
 stopasgroup = true
 
 [program:nginx]
-command=docker-compose up --no-color --exit-code-from nginx-proxy nginx-proxy
+command=docker-compose run --rm --service-ports nginx-proxy
 stdout_events_enabled=true
 stderr_events_enabled=true
-stopsignal = KILL
+stopsignal = INT
 stopasgroup = true
 
 [eventlistener:logger]
```

robertknight replied:
Regarding the `stopsignal` change above, I wonder whether `SIGKILL` may be too aggressive for other processes as well by preventing them from doing cleanup that they otherwise would do.

seanh replied:
I had tried `--abort-on-container-exit` (as Superdev does) but it doesn't do what we want:

```terminal
$ git diff
diff --git a/conf/supervisord-dev.conf b/conf/supervisord-dev.conf
index 12622bb..9eec592 100644
--- a/conf/supervisord-dev.conf
+++ b/conf/supervisord-dev.conf
@@ -10,7 +10,7 @@ stopsignal = KILL
 stopasgroup = true
 
 [program:nginx]
-command=docker-compose up --no-color --exit-code-from nginx-proxy nginx-proxy
+command=docker-compose up --no-color --exit-code-from nginx-proxy --abort-on-container-exit nginx-proxy
 stdout_events_enabled=true
 stderr_events_enabled=true
 stopsignal = KILL
$ # Nothing running:
$ tox -qqe dev --run-command 'docker-compose ps'
Name   Command   State   Ports
------------------------------
~/P/via3 supervisor> make dev
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498512] [INFO] Starting gunicorn 20.0.4
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498512] [INFO] Listening at: http://0.0.0.0:9082 (3498512)
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498512] [INFO] Using worker: sync
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498538] [INFO] Booting worker with pid: 3498538
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498549] [INFO] Booting worker with pid: 3498549
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498567] [INFO] Booting worker with pid: 3498567
nginx (stderr)       | Creating network "via3_default" with the default driver
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498600] [INFO] Booting worker with pid: 3498600
web                  | Cache buster salt: f730d93b5919787dd79832a0244a648a
web                  | Cache buster salt: f730d93b5919787dd79832a0244a648a
nginx (stderr)       | Creating via3_nginx-proxy_1 ... 
web                  | Cache buster salt: f730d93b5919787dd79832a0244a648a

nginx (stderr)       | Creating via3_nginx-proxy_1 ... done
nginx (stderr)       |

nginx                | Attaching to via3_nginx-proxy_1
^C
ERROR: got KeyboardInterrupt signal
130! ~/P/via3 supervisor> # make dev has not killed nginx:
130! ~/P/via3 supervisor> tox -qqe dev --run-command 'docker-compose ps'
       Name                     Command               State                Ports
----------------------------------------------------------------------------------------------
via3_nginx-proxy_1   /bin/sh -c /etc/nginx/dev_ ...   Up      80/tcp, 127.0.0.1:9083->9083/tcp
$ 
```

I think `--abort-on-container-exit` tells Docker Compose to kill the _other containers_ (in Via 3's case, there aren't any other containers) when one container stops ("Stops all containers if any container was stopped"). It doesn't tell Docker Compose to kill the container(s) when the `docker-compose up` command itself is <kbd><kbd>Ctrl</kbd> + <kbd>c</kbd></kbd>'d. It has to do with multi-container Docker Compose situations, doesn't do anything in a single-container situatin like Via 3's

seanh replied:
@robertknight Yes! The `docker-compose run --rm ...` command seems to do what we want as long as it's combined with the `stopsignal = INT` change. It also solves two other problems with docker-compose:

1. It removes "double-enveloping of nginx's log messages because docker-compose prefixes them with `nginx-proxy_1  |` and then supervisor prefixes them again with `nginx                |` so they all come out like `nginx                | nginx-proxy_1  | ...`. With `docker-compose run` Docker Compose no longer prefixes things so they come out as we want: `nginx                | ...`

2. It also gets rid of the problem that `make dev` would start by printing out lots of Nginx logs messages from the _previous_ run of `make dev` so you get an irrelevant screenful of junk. Generally having a lot of stuff get printed out when `make dev` starts if something to avoid because it's annoying and it hides any errors or warnings in all the noise

So this is a great improvement, thanks

seanh replied:
Done

seanh commented on requirements-dev.in:
```diff
@@ -1,6 +1,7 @@
 ipython
 ipdb
+supervisor
+docker-compose
```
To allow supervisor to run docker-compose it was necessary to remove the separate docker-compose tox env and install docker-compose in the dev env that supervisor is run in. This makes sense: NGINX is a dev process not a service (in other apps the separate docker-compose env is used for running services)

jon-betts commented on Makefile:
```diff
@@ -19,21 +21,16 @@ help:
-nginx: args?=up
-nginx: python
-	@tox -qe docker-compose -- $(args)
-
```
When developing Via 3, it is necessary to run `nginx` separately from the web app. This is the same sort of feedback as the last PR in this area, so I won't go over all the reasons again.  But we should not remove a developer convenience, as this PR otherwise makes it strictly worse for me.

seanh replied:
I often want to run the web process (for example) separately to debug it. The way I've been doing this in lms and h is:

```
$ # Look in supervisor-dev.conf for the web process's command
~/P/via3 supervisor> cat conf/supervisord-dev.conf 
[supervisord]
nodaemon = true
silent = true

[program:web]
command=gunicorn -c conf/gunicorn/dev.conf.py --paste conf/development.ini
...

$ # Run the command copy-pasted from supervisor-dev.conf, in tox.
$ tox -qqe dev --run-command 'gunicorn -c conf/gunicorn/dev.conf.py --paste conf/development.ini'
[2020-08-10 13:55:32 +0100] [3577213] [INFO] Starting gunicorn 20.0.4
...
```

It helps if all of our apps use the same commands for the same processes.

Anyway, this has been working well for me and I think it might be the most reasonable solution. The problem with things like `make web` is that, currently on Via 3's master branch:

1. `make dev` would run tox's `dev` environment
2. tox would run `honcho start` (or supervisor, now) inside the dev env
3. honcho (or supervisor) would run `make web` and `make nginx`
4. Now our second instance of make (running inside tox, which is in turn running inside the first instance of make) would run tox again
5. Now our second instance of tox (running inside a make that's running inside a tox that's running inside a make) would run `gunicorn`
6. A third instance of tox (in a make in a tox in a make) would then run `docker-compose` in another tox env

The convolutedness of this makes the dev env hard to understand (what happens when I run `make dev`?) and it also means that `make dev` has a very complex and deep process tree with multiple instances of make and tox running simultaneously and inside of each other.

This appears to have been the cause of multiple issues we've been having (and Via 3 on master is still having) with `make dev` being slow to start up and shut down, messing up terminals after shutting down, leaving behind zombie processes and celery pid files etc, crashing when trying to autoreload the web server when the source code changes, etc.

It's difficult to pinpoint all or any of the above problems to one exact thing. They likely don't even have a single cause. One thing I can say is that they were _not_ caused by honcho: I was still able to reproduce the issues after moving from honcho to supervisor. So the general approach I took was "boil `make dev` down to the simplest possible process tree" and avoid running multiple instances of make or tox. Based on the last few weeks of problems this seems to have fixed the issues for h and lms.

jon-betts commented on bin/logger:
```diff
@@ -90,20 +135,36 @@ def _write(s):
 
-if __name__ == '__main__':
+if __name__ == "__main__":
     main()
```
This is a lot of boilerplate code. If we were to spot a mistake in it, it would involve a lot of identikit PRs to fix it. Or more worrying, we might diverge / not bother fixing it.

jon-betts reviewed (changes requested):
I've not gone through testing this locally yet, as there is some feedback we I think can be addressed first.

 * The major one, which is similar to a number of other PR's in this area is that two independently runnable commands have now become one. We've discussed this a number of times, so I won't rehash the reasons. Just to say the requirement for running them separately remains even if technology changes.
 * The issue with docker-compose not exiting can be solved using the method `superdev` does, which is linked below.

The logging script is getting quite long. This is more of a long term thing, as this isn't the straw that broke the camels back or anything. This is quite a lot of boilerplate, and will likely cause a lot of PRs if we ever have trouble with it. It might be worth thinking about some mechanism to centralise or share this.

robertknight commented on conf/supervisord-dev.conf:
```diff
@@ -0,0 +1,35 @@
+file = .supervisor.sock
+
+[rpcinterface:supervisor]
+supervisor.rpcinterface_factory = supervisor.rpcinterface:make_main_rpcinterface
```
I noticed that h's `conf/supervisord-dev.conf` doesn't have this section. Should it?

seanh replied:
Yes, it should. This section is for the `make supervisor` command which h doesn't have. It's actually not a super useful command though (I haven't found myself using it) so not a high priority. We could possibly even remove it from Via 3 and LMS

robertknight reviewed (commented):
I tested this manually and for simply running the application in development it worked as expected, with the exception of the `make dev` issue. I proposed a slightly different solution to what Jon suggested. I'm not familiar enough with `docker-compose` to fully understand all the differences.

seanh committed:
Use `docker-compose run` to run nginx

Fixes https://github.com/hypothesis/via3/issues/221

While this doesn't remove Docker Compose (as #221 suggests) it does fix
all of the problems described in #221.

seanh commented:
> Regarding the stopsignal change above, I wonder whether SIGKILL may be too aggressive for other processes as well by preventing them from doing cleanup that they otherwise would do.

Let's keep an eye on this. If we notice any problems that look like they may be related to SIGKILL, this might be worth trying

seanh committed:
Add `make web` and `make nginx`

This introduces convenience commands for running just the Pyramid app or
just the NGINX component, at the cost of duplicating the commands in
Makefile and supervisord-dev.conf.

jon-betts reviewed (approved):
I'm not sure what the differences are between how we run the services here vs. superdev, but practically it seems to work out the same. The logging issue has also been fixed from what I can see.

The colouration is nice and the services seem to start, and shut down clean. The separate commands work well too. :+1:
